redis/src/cluster_async/mod.rs

```diff
@@ -377,17 +380,67 @@ struct RequestInfo<C> {
 impl<C> RequestInfo<C> {
     fn set_redirect(&mut self, redirect: Option<Redirect>) {
         if let Some(redirect) = redirect {
-            let redirect = InternalSingleNodeRouting::Redirect(redirect);
             match &mut self.cmd {
-                CmdArg::Cmd { routing, .. } => {
-                    *routing = redirect.into();
-                }
+                CmdArg::Cmd { routing, .. } => match routing {
+                    InternalRoutingInfo::SingleNode(route) => {
+                        let redirect = InternalSingleNodeRouting::Redirect {
+                            redirect,
+                            previous_routing: Box::new(std::mem::replace(
+                                route,
+                                InternalSingleNodeRouting::Random,
+                            )),
+                        }
+                        .into();
+                        *routing = redirect;
+                    }
+                    InternalRoutingInfo::MultiNode(_) => {
+                        panic!("Cannot redirect multinode requests")
+                    }
+                },
                 CmdArg::Pipeline { route, .. } => {
+                    let redirect = InternalSingleNodeRouting::Redirect {
+                        redirect,
+                        previous_routing: Box::new(std::mem::replace(
+                            route,
+                            InternalSingleNodeRouting::Random,
+                        )),
+                    };
                     *route = redirect;
                 }
             }
         }
     }
+
+    fn reset_redirect(&mut self) {
+        match &mut self.cmd {
+            CmdArg::Cmd { routing, .. } => {
+                if let InternalRoutingInfo::SingleNode(InternalSingleNodeRouting::Redirect {
+                    previous_routing,
+                    ..
+                }) = routing
+                {
+                    let mut previous_routing = std::mem::replace(
```

Comment:
still can't understand :( the var names are confusing too.
you take previous_routing_1 and put Random in it, then previous_routing_2 takes the old value of previous_routing_1.
so now
previous_routing_2 = previous_routing_1
previous_routing_1 = Random
and then you swap routing, which contains a reference to previous_routing_1, and replace it with previous_routing_2?
help.